### PR TITLE
feat(core): add passwordless register flow of phone

### PR DESCRIPTION
**Pull request by darcyYe**
## Summary

Add passwordless register flow (by phone number and SMS).

## Details
If `phone number` is provided but not `code`, check for `phone number` existence and send passcode via SMS if not.
Check the validity of `phone number` and `code` if both of them are provided, create a new user if match.

## Linear Issue Reference

LOG-849

## Testing

1. phone number exists
image

2. `phone number` provided but `code` is not, passcode should be send to provided phone number
image
image

3. `phone number` and wrong `code` are provided
image

4. `phone number` and right `code` are provided, per screenshot, a new account was successfully created
image
image

### CODE REVIEW

**Comment by wangsijie** [comment]
You can try `<details>` in PR description. 😺

File: packages/core/src/lib/register.ts [hunk 1]
```diff
@@ -58,6 +58,9 @@ export const registerWithUsernameAndPassword = async (
   });
 
   await assignRegistrationResult(ctx, provider, id);
+  ctx.userLog.userId = id;
```
Comment (wangsijie) [inline packages/core/src/lib/register.ts:61]: currently we don't have registration logs, we can handle this later in https://linear.app/silverhand/project/core-logs-451d2cd0dc86

**Review by wangsijie — CHANGES_REQUESTED** [review]

File: packages/core/src/lib/register.ts (hunk 1 above)
Comment (darcyYe) [reply packages/core/src/lib/register.ts:61]: an issue added for user registration log ([LOG-1512](https://linear.app/silverhand/issue/LOG-1512/log-register)). these user logs can help us to make more sense of users' preferences. what do you think? @logto-io/eng

**Review by gao-sun — CHANGES_REQUESTED** [review]
The scope of PR title is wrong. Should be `feat(core)`

File: packages/core/src/lib/register.ts [hunk 2]
```diff
@@ -96,4 +114,33 @@ export const registerWithEmailAndPasscode = async (
...
```
[the remark this answers is not in the document]
Comment (darcyYe) [reply packages/core/src/lib/register.ts:117]: Linear issue added (LOG-1516) as a reminder. Will refactor this when fix user logs functionality.

File: packages/core/src/lib/register.ts [hunk 3]
```diff
@@ -75,6 +78,21 @@ export const sendPasscodeToEmail = async (ctx: Context, jti: string, email: stri
   ctx.state = 204;
 };
 
+export const sendPasscodeToPhone = async (ctx: Context, jti: string, phone: string) => {
+  assertThat(phoneRegEx.test(phone), new RequestError('user.invalid_phone'));
+  assertThat(
+    !(await hasUserWithPhone(phone)),
+    new RequestError({
+      code: 'user.phone_exists',
```
[the remark this answers is not in the document]
Comment (darcyYe) [reply packages/core/src/lib/register.ts]: error description `username/phone/email_exists_register` added.
should we also keep `username/phone/email_exists` for other case when existence check is needed? for example, user register with phone number and is trying to add phone number to his/her account.

File: packages/core/src/lib/register.ts (hunk 1 above)
Comment (wangsijie) [reply packages/core/src/lib/register.ts:61]: you are right, but user logs is beyond current project.

**Review by wangsijie — APPROVED** [review]

File: packages/core/src/lib/register.ts (hunk 1 above)
Comment (darcyYe) [reply packages/core/src/lib/register.ts:61]: let's keep this as a reminder and fix user logs functionality later.

File: packages/core/src/lib/register.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (gao-sun) [reply packages/core/src/lib/register.ts]: let's refactor for other cases until when we need

File: packages/phrases/src/locales/zh-cn.ts [hunk 4]
```diff
@@ -47,6 +47,10 @@ const errors = {
   user: {
     username_exists: '用户名已存在。',
     email_exists: '邮箱地址已存在。',
+    phone_exists: '手机号码已存在。',
```
Comment (gao-sun) [inline packages/phrases/src/locales/zh-cn.ts:50]: we can remove this until needed. also `email_exists` can be optimized in a separate PR

File: packages/phrases/src/locales/en.ts [hunk 5]
```diff
@@ -46,6 +46,12 @@ const errors = {
   user: {
     username_exists: 'The username already exists.',
     email_exists: 'The email already exists.',
+    phone_exists: 'The phone already exists.',
+    username_exists_register: 'The username can not be regitsered because it is occupied.',
+    email_exists_register:
+      'The email address can not be regitsered because it is bound to another accournt.',
```
Comment (gao-sun) [inline packages/phrases/src/locales/en.ts:52]: `The email address has been registered.`? also for `phone` and `username`

**Review by gao-sun — APPROVED** [review]

File: packages/phrases/src/locales/zh-cn.ts (hunk 4 above)
Comment (darcyYe) [reply packages/phrases/src/locales/zh-cn.ts]: removed it for now.

File: packages/phrases/src/locales/en.ts (hunk 5 above)
Comment (darcyYe) [reply packages/phrases/src/locales/en.ts]: fixed.
